be/src/storage/local_primary_key_recover.cpp

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/local_primary_key_recover.h"
+
+#include "storage/chunk_helper.h"
+#include "storage/tablet_meta_manager.h"
+#include "storage/update_manager.h"
+
+namespace starrocks {
+
+Status LocalPrimaryKeyRecover::pre_cleanup() {
+    // remove pk index
+    auto index_entry = _update_mgr->index_cache().get(_tablet->tablet_id());
+    if (index_entry != nullptr) {
+        _update_mgr->index_cache().remove(index_entry);
+    }
+    // delete persistent index meta
+    if (_tablet->get_enable_persistent_index()) {
+        RETURN_IF_ERROR(TabletMetaManager::clear_persistent_index(_tablet->data_dir(), &_wb, _tablet->tablet_id()));
+    }
+    // Notice. We don't remove delvec here, otherwise replace delvec by latest apply version later
+    return Status::OK();
+}
+
+starrocks::Schema LocalPrimaryKeyRecover::generate_pkey_schema() {
+    const TabletSchemaCSPtr tablet_schema_ptr = _tablet->tablet_schema();
+    vector<ColumnId> pk_columns(tablet_schema_ptr->num_key_columns());
+    for (auto i = 0; i < tablet_schema_ptr->num_key_columns(); i++) {
+        pk_columns[i] = (ColumnId)i;
+    }
+    return ChunkHelper::convert_schema(tablet_schema_ptr, pk_columns);
+}
+
+Status LocalPrimaryKeyRecover::sort_rowsets(std::vector<RowsetSharedPtr>* rowsets) {
+    std::sort(rowsets->begin(), rowsets->end(), [](const RowsetSharedPtr& a, const RowsetSharedPtr& b) {
+        const RowsetMetaSharedPtr& rowset_meta_a = a->rowset_meta();
+        const RowsetMetaSharedPtr& rowset_meta_b = b->rowset_meta();
+        // if rowset was generated by compaction, use max compact input rowset id as its compare id
+        uint32_t rowset_a_comp_id = rowset_meta_a->has_max_compact_input_rowset_id()
+                                            ? rowset_meta_a->max_compact_input_rowset_id()
+                                            : rowset_meta_a->get_rowset_seg_id();
+        uint32_t rowset_b_comp_id = rowset_meta_b->has_max_compact_input_rowset_id()
+                                            ? rowset_meta_b->max_compact_input_rowset_id()
+                                            : rowset_meta_b->get_rowset_seg_id();
```

Comment:
When we upgrade from old version and `max_compact_input_rowset_id` does not exist, is it correct to use `rowset_seg_id` as the basis for sorting rowsets?